Allow to pass named arguments to macro calls

GuillaumeGomez opened the pull request:
Fixes #885.

Based on https://github.com/djc/askama/pull/909 (so only the last 3 commits are relevant to this PR).

I made a very permissive implementation which allows to mix named and non-named arguments. The only thing we need to check outside of the parser is that the named arguments actually have an equivalent in the called macro. For the rest, everything is checked in the parser directly, which allows nicer error messages.

GuillaumeGomez commented:
Rebased.

djc commented on askama_derive/src/generator.rs:
```diff
@@ -727,6 +727,42 @@ impl<'a> Generator<'a> {
+        {
+            // First we check that all named arguments actually exist in the called item.
+            for arg in args.iter() {
+                if let Expr::NamedArgument(arg_name, _) = arg {
```
Let's do `let .. else` for the outer branch here?

GuillaumeGomez replied:
I don't really see how it would improve the code since there is no `else` clause here?

djc replied:
It reduces rightward drift for the further check, so I would still like to do this.

GuillaumeGomez replied:
Now there is a good reason for it (the `break` which was added only recently) so adding it!

djc commented on askama_derive/src/generator.rs:
```diff
@@ -727,6 +727,42 @@ impl<'a> Generator<'a> {
                 args.len()
             )));
         }
+        let mut output_args = Vec::new();
```
This seems pretty messy. I'd like to have a comment here that explains the high-level logic for handling invocations that have both named and unnamed arguments, which seems like the tricky case here. Then, I think there should be one pass over the arguments that both splits the arguments in `unnamed_args` and `named_args` (I think we should be able to do this without cloning?) and checks whether all the names are acceptable.

djc replied:
This is still iterating `args` twice: once for `args.iter().any()` and once for `args.iter().partition()`. I'm pretty sure this isn't actually necessary?

GuillaumeGomez replied:
I can simplify it even further indeed. I'll do that.

GuillaumeGomez replied:
My idea couldn't work actually (sorry for the noise). The problem is that I first need to **ensure** that all named arguments actually have an equivalent in the macro, otherwise they'll be passed as non-named arguments and the `remove` call would fail. However, a first improvement that could be done would be to move this loop after we split both kind of arguments so that we only iterate on named arguments.

GuillaumeGomez replied:
Since named arguments are always last, it allowed to remove the first loop and to instead just check if the last element is a named argument.

djc commented on askama_parser/src/expr.rs:
```diff
@@ -74,15 +77,45 @@ pub enum Expr<'a> {
 impl<'a> Expr<'a> {
     pub(super) fn arguments(i: &'a str, level: Level) -> ParseResult<'a, Vec<Self>> {
         let (_, level) = level.nest(i)?;
+        let mut named_arguments = HashSet::new();
```
I don't think this logic should live in `Expr::arguments()`, which is not just used for macros.

GuillaumeGomez replied:
It allows to greatly improve the parser error, but as you prefer.

GuillaumeGomez replied:
I updated the code to not always allow to have named arguments. It'll fail earlier in this case.

However, I still think we should keep the `named_arguments` set checks for duplicated names as it makes the errors much better (otherwise we don't get the origin of the error since we don't have the template text anymore when generating the code).

djc commented on book/src/template_syntax.md:
```diff
@@ -587,3 +587,49 @@ You can optionally specify the name of the macro in `endmacro`:
+
+{% endmacro %}
+
+{% call heading(bold="something", arg = "title") %}
```
I think the documentation should be consistent in whether it uses spacing around `=` even though the parser should cope with both.

GuillaumeGomez replied:
Updated! I also precised that we can have whitespace characters around `=`.

GuillaumeGomez commented:
Is there anything else to be done here?

djc commented on askama_parser/src/expr.rs:
```diff
@@ -72,17 +75,74 @@ pub enum Expr<'a> {
         )(i)
     }
 
+    fn arguments_parsing(
```
I'd prefer to inline this. I wonder if we can move the `if` inside the `alt` somehow?

```rust
alt((
    match is_jinja_macro {
        true => move |i| Self::named_argument(i, level, named_arguments, start),
        false => move |i| nom::err::Error::something(), // TODO?
    },
    move |i| Self::parse(i, level),
))(i)
```

GuillaumeGomez replied:
We can actually do even better and move the `is_macro` check directly into `named_argument` to make it return an error if it's not the case. Much simpler than currently.

djc replied:
I still think we should inline `arguments_parsing()`.

GuillaumeGomez replied:
I find the code a bit less nice to follow but if you prefer it inlined, let's do it! :)

djc commented on askama_parser/src/expr.rs:
```diff
@@ -72,17 +75,74 @@ pub enum Expr<'a> {
+    pub(super) fn arguments(
+        i: &'a str,
+        level: Level,
+        is_jinja_macro: bool,
```
Prefer to rename this to `is_template_macro`, since this isn't Jinja.

GuillaumeGomez replied:
:+1:

djc commented on askama_parser/src/expr.rs:
```diff
@@ -72,17 +75,74 @@ pub enum Expr<'a> {
+        let (_, level) = level.nest(i)?;
+        let (i, (argument, _, value)) =
+            tuple((identifier, ws(char('=')), move |i| Self::parse(i, level)))(i)?;
+        if !named_arguments.insert(argument) {
```
Please invert this `if` with an early return so that we can avoid rightward drift for the error case.

djc commented on askama_derive/src/generator.rs:
```diff
@@ -727,6 +727,52 @@ impl<'a> Generator<'a> {
                 args.len()
             )));
         }
+        // Handling both named and unnamed arguments requires to be careful of the arguments order.
```
I think I'd prefer a more restrictive way of combining named and unnamed arguments, which is that we only allow named arguments after positional arguments.

djc replied:
Please update this comment to be in line with the updated logic.

GuillaumeGomez commented:
Updated! I applied your suggestions and also enforced that named arguments should always be the last arguments. As mentioned above, it allowed to remove a loop since we can check if we have named arguments by checking if the last argument is a named one.

djc commented on askama_derive/src/generator.rs:
```diff
@@ -759,6 +749,18 @@ impl<'a> Generator<'a> {
             let (mut named_arguments, mut non_named_arguments): (Vec<_>, Vec<_>) = args
                 .iter()
                 .partition(|item| matches!(item, Expr::NamedArgument(_, _)));
+
```
These changes should be inlined into the original commits.

GuillaumeGomez commented:
Squashed commits so named arguments coming last are now part of the first commit.

djc commented on askama_derive/src/generator.rs:
```diff
@@ -727,6 +727,57 @@ impl<'a> Generator<'a> {
+        let mut output_args = Vec::new();
+        // Since named arguments can only be passed last, we only need to check if the last argument
+        // is a named one.
+        let args = if args
```
I think this can be simplified to `if let Some(Expr::NamedArgument(_, _)) = args.last()`?

GuillaumeGomez replied:
Much better indeed!

djc commented on askama_derive/src/generator.rs:
```diff
@@ -727,6 +727,57 @@ impl<'a> Generator<'a> {
+                .partition(|item| matches!(item, Expr::NamedArgument(_, _)));
+
+            // First we check that all named arguments actually exist in the called item.
+            for arg in named_arguments.iter() {
```
I think the right logic here is:

- Set up a `named_args` `HashMap<&str, &Expr>`
- For each argument, in reverse order:
  - If the argument is not named, break out of the loop
  - If it is named, insert into `named_args`
- Then, change the `iter::zip()` loop to add `.enumerate()`
  - If an argument is named, use the `named_args` map to lookup the actual argument

GuillaumeGomez commented:
The idea of using a `HashMap` was really good. I simplified the code quite a lot: I loop through the template arguments, for each of them, I check if there is a named argument with this name in the `HashMap` and if so use the associated `Expr`.

With this, we avoid allocating other `Vec`s and remove a few loops.

I updated the code comments as well.

GuillaumeGomez commented on askama_derive/src/generator.rs:
```diff
@@ -727,7 +727,41 @@ impl<'a> Generator<'a> {
+        // * If there is one, we add it and move to the next argument.
+        // * If there isn't one, then we pick the next argument (we can do it without checking
+        //   anything since named arguments are always last).
+        let mut index = 0;
```
Instead of this index, I could use an iterator. As you prefer.

djc replied:
This seems fine.

djc commented on book/src/template_syntax.md:
```diff
@@ -587,3 +587,57 @@ You can optionally specify the name of the macro in `endmacro`:
+{% macro heading(arg1, arg2, arg3, arg4) %}
+{% endmacro %}
+
+{% call heading("something", "b", arg4="ah", arg2="title") %}
```
Ahh, sorry, these aren't quite the semantics I was going for. The positional arguments should always retain their position, so it should not be allowed to specify `arg2` by name here since the call also passes two positional arguments.

GuillaumeGomez replied:
That however seems a bit too much of a restriction imo. We can't mix named and unnamed arguments, so I would at least want to be able to specify some arguments by name (because it's not obvious what they are just from the value) whether or not it is one of the first arguments without having to specify all arguments as named ones. If that makes sense to you?

djc commented on askama_derive/src/generator.rs:
```diff
@@ -727,7 +727,50 @@ impl<'a> Generator<'a> {
+        // * If there is one, we add it and move to the next argument.
+        // * If there isn't one, then we pick the next argument (we can do it without checking
+        //   anything since named arguments are always last).
+        let mut args_iterator = args
```
The `args_iterator` approach is unnecessarily complex. We can just have an `allow_positional` that is initialized to `true` which is changed to `false` on the first occurrence of a named argument.

djc replied:
How less good? Do you have an example?

GuillaumeGomez replied:
Before I could say `it's bad to use named argument "x" here`. Now I simply provide the argument name of the template macro. I could maybe provide the index too.

But I have to admit, before it was a bit strange too:

> is taking the position of an unnamed argument

Not very clear.

GuillaumeGomez commented:
Updated!

djc reviewed (approved):
Okay, thanks for sticking with this, I think this is in a pretty good shape now!

GuillaumeGomez commented:
Thanks for all the review rounds! I'll continue my docs.rs migration and come back here if anything else is missing. :)

djc commented:
Sounds great!
